File: problems/803_bricks_falling_when_hit/Solution.java
class Solution {
    int[] parent, size;
    public int[] hitBricks(int[][] grid, int[][] hits) {
        int m = grid.length, n = grid[0].length;
        parent = new int[m * n + 1];
        size = new int[m * n + 1];
        for (int i = 0; i <= m * n; i++) { parent[i] = i; size[i] = 1; }
        int TOP = m * n;
        int[][] gridCopy = new int[m][n];
        for (int i = 0; i < m; i++) gridCopy[i] = grid[i].clone();
        for (int[] h : hits) gridCopy[h[0]][h[1]] = 0;
        for (int c = 0; c < n; c++) if (gridCopy[0][c] == 1) union(c, TOP);
        for (int r = 1; r < m; r++) for (int c = 0; c < n; c++) {
            if (gridCopy[r][c] == 0) continue;
            if (gridCopy[r - 1][c] == 1) union(r * n + c, (r - 1) * n + c);
            if (c > 0 && gridCopy[r][c - 1] == 1) union(r * n + c, r * n + c - 1);
        }
        int[] result = new int[hits.length];
        int[][] dirs = {{-1,0},{1,0},{0,-1},{0,1}};
        for (int k = hits.length - 1; k >= 0; k--) {
            int r = hits[k][0], c = hits[k][1];
            if (grid[r][c] == 0) continue;
            int prev = size[find(TOP)];
            if (r == 0) union(r * n + c, TOP);
            for (int[] d : dirs) {
                int nr = r + d[0], nc = c + d[1];
                if (nr >= 0 && nr < m && nc >= 0 && nc < n && gridCopy[nr][nc] == 1)
                    union(r * n + c, nr * n + nc);
            }
            gridCopy[r][c] = 1;
            result[k] = Math.max(0, size[find(TOP)] - prev - 1);
        }
        return result;
    }
    int find(int x) { while (parent[x] != x) { parent[x] = parent[parent[x]]; x = parent[x]; } return x; }
    void union(int x, int y) {
        int px = find(x), py = find(y);
        if (px == py) return;
        if (size[px] < size[py]) { int t = px; px = py; py = t; }
        parent[py] = px; size[px] += size[py];
    }
}

File: problems/803_bricks_falling_when_hit/solution.py
from typing import List


class Solution:
    def hitBricks(self, grid: List[List[int]], hits: List[List[int]]) -> List[int]:
        m, n = len(grid), len(grid[0])
        parent = list(range(m * n + 1))
        size = [1] * (m * n + 1)
        def find(x: int) -> int:
            while parent[x] != x:
                parent[x] = parent[parent[x]]
                x = parent[x]
            return x
        def union(x: int, y: int):
            px, py = find(x), find(y)
            if px == py: return
            if size[px] < size[py]: px, py = py, px
            parent[py] = px
            size[px] += size[py]
        def index(r: int, c: int) -> int:
            return r * n + c
        TOP = m * n
        grid_copy = [row[:] for row in grid]
        for r, c in hits:
            grid_copy[r][c] = 0
        for c in range(n):
            if grid_copy[0][c] == 1:
                union(index(0, c), TOP)
        for r in range(1, m):
            for c in range(n):
                if grid_copy[r][c] == 0: continue
                if grid_copy[r - 1][c] == 1:
                    union(index(r, c), index(r - 1, c))
                if c > 0 and grid_copy[r][c - 1] == 1:
                    union(index(r, c), index(r, c - 1))
        result = [0] * len(hits)
        for k in range(len(hits) - 1, -1, -1):
            r, c = hits[k]
            if grid[r][c] == 0:
                continue
            prev = size[find(TOP)]
            if r == 0:
                union(index(r, c), TOP)
            for dr, dc in [(-1, 0), (1, 0), (0, -1), (0, 1)]:
                nr, nc = r + dr, c + dc
                if 0 <= nr < m and 0 <= nc < n and grid_copy[nr][nc] == 1:
                    union(index(r, c), index(nr, nc))
            grid_copy[r][c] = 1
            result[k] = max(0, size[find(TOP)] - prev - 1)
        return result
